django/views/i18n.py

```diff
@@ -37,6 +42,23 @@ def set_language(request):
     if request.method == 'POST':
         lang_code = request.POST.get('language', None)
         if lang_code and check_for_language(lang_code):
+            # Try to convert the `next` url into a translated url
+            parsed = urlsplit(next)
+            try:
+                match = resolve(parsed.path)
+            except Resolver404:
+                pass
+            else:
+                activate(lang_code)
+                to_be_reversed = "%s:%s" % (match.namespace, match.url_name) if match.namespace else match.url_name
```

Comment:
are both branches tested here?
Do you think factoring at least part of the new logic into a `urlresolvers.py` function might be useful? These additions make the view seem quite a bit more complex.